src/digital_ocean.rs

```diff
@@ -74,19 +74,38 @@ impl DigitalOceanClient {
                 let ip_address_array = droplet_json["networks"]["v4"].as_array().ok_or(
                     Error::MalformedDigitalOceanApiRespose("droplets".to_string()),
                 )?;
-                let public_ip = ip_address_array
-                    .iter()
-                    .find(|x| x["type"].as_str().unwrap() == "public")
-                    .ok_or(Error::DigitalOceanPublicIpAddressNotFound)?;
-                droplets.push(Droplet {
-                    id: id as usize,
-                    name,
-                    ip_address: Ipv4Addr::from_str(
+                // The following might fail if we start multiple networks in parallel.
+                let get_ip_address = || -> Result<Ipv4Addr, Error> {
+                    let public_ip = ip_address_array
+                        .iter()
+                        .find(|x| x["type"].as_str().unwrap() == "public")
+                        .ok_or(Error::DigitalOceanPublicIpAddressNotFound)?;
+
+                    let ip_address = Ipv4Addr::from_str(
                         public_ip["ip_address"]
                             .as_str()
                             .ok_or(Error::DigitalOceanPublicIpAddressNotFound)?,
-                    )?,
-                });
+                    )?;
+
+                    Ok(ip_address)
+                };
+
+                match get_ip_address() {
```

Comment:
Not sure a closure is needed here

```suggestion
                let get_ip_address = {
                    let public_ip = ip_address_array
                        .iter()
                        .find(|x| x["type"].as_str().unwrap() == "public")
                        .ok_or(Error::DigitalOceanPublicIpAddressNotFound)?;

                    let ip_address = Ipv4Addr::from_str(
                        public_ip["ip_address"]
                            .as_str()
                            .ok_or(Error::DigitalOceanPublicIpAddressNotFound)?,
                    )?;

                    Ok(ip_address)
                };

                match get_ip_address {
```

Reply:
I'm not sure if that is possible? The `?` operator only propagates the error out of a function or a closure. Here we try to capture it using a standard `{ }` and it would not be possible.

Reply:
You're right! Hope [`try`](https://doc.rust-lang.org/beta/unstable-book/language-features/try-blocks.html) is stabilised one day.

Reply:
That is quite nice to have!